File: C++/shakerSort.cpp
#include <iostream>
#include <vector>

// Function to implement Shaker Sort
void shakerSort(std::vector<int>& arr) {
    int n = arr.size(); // Get the size of the array
    bool swapped = true; // Flag to indicate if any swaps were made
    int start = 0;       // Starting index for the current pass
    int end = n - 1;     // Ending index for the current pass

    while (swapped) {
        swapped = false; // Reset the swapped flag for the current pass

        // Forward pass (left to right)
        for (int i = start; i < end; ++i) {
            if (arr[i] > arr[i + 1]) {
                std::swap(arr[i], arr[i + 1]); // Swap if the current element is greater than the next
                swapped = true;                 // Set the flag indicating a swap occurred
            }
        }

        // If no swaps were made, the array is sorted
        if (!swapped) {
            break;
        }

        // Reduce the end index since the last element is now in its correct position
        end--;

        // Reset the swapped flag for the backward pass
        swapped = false;

        // Backward pass (right to left)
        for (int i = end; i > start; --i) {
            if (arr[i] < arr[i - 1]) {
                std::swap(arr[i], arr[i - 1]); // Swap if the current element is less than the previous
                swapped = true;                 // Set the flag indicating a swap occurred
            }
        }

        // Increase the start index since the first element is now in its correct position
        start++;
    }
}

// Helper function to print the elements of the array
void printArray(const std::vector<int>& arr) {
    for (int i = 0; i < arr.size(); ++i) {
        std::cout << arr[i] << " "; // Print each element followed by a space
    }
    std::cout << std::endl; // Print a newline at the end
}

// Main function
int main() {
    // Example array to sort
    std::vector<int> arr = {5, 3, 8, 4, 2};

    // Output the original array
    std::cout << "Original array: ";
    printArray(arr);

    // Call the shaker sort function
    shakerSort(arr);

    // Output the sorted array
    std::cout << "Sorted array: ";
    printArray(arr);

    return 0;
}

File: Python/shakerSort.py
def shaker_sort(arr):
    
    n = len(arr)  # Get the length of the array
    # Initialize the start and end pointers for the current pass
    start = 0
    end = n - 1
    swapped = True  # Flag to check if any swapping occurred

    while swapped:
        swapped = False
        
        # Traverse the array from left to right
        for i in range(start, end):
            if arr[i] > arr[i + 1]:  # If the current element is greater than the next
                arr[i], arr[i + 1] = arr[i + 1], arr[i]  # Swap them
                swapped = True  # Mark that a swap occurred
        
        # If nothing was swapped, the array is sorted
        if not swapped:
            break

        # We can reduce the end pointer since the last element is in place
        end -= 1
        
        swapped = False  # Reset the swapped flag for the next pass
        
        # Traverse the array from right to left
        for i in range(end, start, -1):
            if arr[i] < arr[i - 1]:  # If the current element is less than the previous
                arr[i], arr[i - 1] = arr[i - 1], arr[i]  # Swap them
                swapped = True  # Mark that a swap occurred
        
        # We can increase the start pointer since the first element is in place
        start += 1

    return arr  # Return the sorted array

# Example usage
if __name__ == "__main__":
    # Sample array to sort
    sample_array = [5, 1, 4, 2, 8, 3]
    print("Original array:", sample_array)
    
    # Sorting the array
    sorted_array = shaker_sort(sample_array)
    print("Sorted array:", sorted_array)
